unused_deps/unused_deps.go

```diff
@@ -116,9 +117,54 @@ func inputFileName(blazeBin, pkg, ruleName, extension string) string {
 	return fmt.Sprintf("%s/%s/%s.%s", blazeBin, pkg, ruleName, extension) // *_{binary,test}
 }
 
-// directDepParams returns --direct_dependency entries from paramsFileName (a jar-2.params file)
-// as a map from jar files to labels.
-func directDepParams(paramsFileNames ...string) (depsByJar map[string]string) {
+// targetLabel reads the manifest of the jar specified by jarFileName.  If the
+// manifest has a "Target-Label" property, targetLabel returns its value.
+func targetLabel(jarFileName string) (string, error) {
+	const BAZEL_TARGET_KEY = "Target-Label:"
+
+	r, err := zip.OpenReader(jarFileName)
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	for _, f := range r.File {
+		if f.Name != "META-INF/MANIFEST.MF" {
+			continue
+		}
+
+		rc, err := f.Open()
+		if err != nil {
+			return "", err
+		}
+		defer rc.Close()
+
+		contents, err := ioutil.ReadAll(rc)
+		if err != nil {
+			return "", err
+		}
+		key := []byte(BAZEL_TARGET_KEY)
+		found := bytes.Index(contents, key)
```

Comment:
I guess the idea with using bytes.Index is to avoid converting all of the bytes from the rc file to string?
It just seems a little obfuscated and you've already read all of the bytes into memory

option 1) change to 'b, err := ioutil.ReadAll(rc) ...' then 'contents := string(b)'
and do all of the rest on the string

option 2) use a bufio.Scanner on rc to avoid having everything in memory at the same time.  Skip until start, then read/copy until \n

Reply:
Sorry about the obfuscation.  I haven't written much Go code recently ... forgot that strings are not encoded.  Thanks very much for reviewing.

It turns out that my code for parsing the value of "Target-Label" did not take into consideration continuation lines, so it truncated long values.  I have fixed that problem in the revised commit I uploaded.